File: api/app/core/schemas/tiktok_downloader.py
from pydantic import BaseModel, constr
from typing import List

from app.core.schemas.camel_case import CamelModel
from app.v1.constants import RegexPatterns


class TikTokVideoDownloaderSchema(BaseModel):
    """
    Represents the schema for a TikTok video downloader request.

    Attributes:
        url (constr): The TikTok video downloader URL.
    """

    url: constr(regex=RegexPatterns.TIKTOK_URL_REGEX)


class TikTokVideoMedia(CamelModel):
    """
    Represents TikTok video media.

    Attributes:
        format_id (str): The ID of the format.
        format_note (str): The format note.
        vcodec (str): The video codec.
        url (str): The URL of the video media.
        resolution (str): The resolution of the video media.
    """

    format_id: str
    format_note: str
    vcodec: str
    url: str
    resolution: str


class TikTokVideoDownloaderResult(CamelModel):
    """
    Represents the result of the TikTok video downloader.

    Attributes:
        medias (List[TikTokVideoMedia]): The list of media for the video.
        creator (str): The creator of the video.
        uploader (str): The uploader of the video.
        title (str): The title of the video.
        description (str): The description of the video.
        duration_string (str): The duration of the video.
        media_type (str): The type of media in the video.
        thumbnail (str): The URL of the video thumbnail.
        format (str): The format of the video.
    """

    medias: List[TikTokVideoMedia] = []
    creator: str
    uploader: str
    title: str
    description: str
    duration_string: str
    media_type: str
    thumbnail: str
    format: str


class TikTokVideoDownloaderResponse(CamelModel):
    """
    A response model for the TikTok video downloader.

    Attributes:
        result (TikTokVideoDownloaderResult): The result of the video downloader.
    """

    result: TikTokVideoDownloaderResult

File: web/src/types/tiktok-downloader.ts
/** Interface for TikTokMedia */
export interface TikTokMedia {
    /** The format id of the media */
    formatId: string 

    /** The format note of the media */
    formatNote: string

    /** The vcodec of the media */
    vcodec: string

    /** The url of the media */
    url: string

    /** The resolution of the media */
    resolution: string
}

/** Interface for TikTokMeta */
export interface TikTokMeta {
    /** The name of the creator */
    creator: string

    /** The name of the uploader */
    uploader: string

    /** The title of the video */
    title: string

    /** The description of the video */
    description: string

    /** The duration of the video in string format */
    durationString: string

    /** The type of the video */
    type: string

    /** The thumbnail of the video */
    thumbnail: string

    /** The format of the video */
    format: string

    /** An error message if there is any */
    error: string
}

/** Interface for TikTokDownloaderResult */
export interface TikTokDownloaderResult extends TikTokMeta {
    /** Array of the media data of the video */
    medias: TikTokMedia[]
}

/** Interface for TikTokDownloaderResponse */
export interface TikTokDownloaderResponse {
    /** The result of the TikTok downloader */
    result: TikTokDownloaderResult
}
